crates/alloy-dyn-contract/src/call.rs

```diff
@@ -0,0 +1,187 @@
+use crate::error::{Error, Result};
+use alloy_dyn_abi::{DynSolValue, FunctionExt};
+use alloy_json_abi::Function;
+use alloy_primitives::{Address, Bytes, U256, U64};
+use alloy_providers::provider::TempProvider;
+use alloy_rpc_types::{state::StateOverride, BlockId, CallInput, CallRequest};
+use std::{
+    future::{Future, IntoFuture},
+    pin::Pin,
+};
+
+/// A builder for sending a transaction via. `eth_sendTransaction`, or calling a function via
+/// `eth_call`.
+///
+/// The builder can be `.await`ed directly which is equivalent to invoking [`CallBuilder::call`].
+///
+/// # Note
+///
+/// Sets the [state overrides](https://geth.ethereum.org/docs/rpc/ns-eth#3-object---state-override-set) for `eth_call`, but this is not supported by all clients.
+pub struct CallBuilder<P> {
+    // todo: this will not work with `send_transaction` and does not differentiate between EIP-1559
+    // and legacy tx
+    request: CallRequest,
+    // todo: only used to decode - should it be some type D to dedupe with `sol!` contracts?
+    function: Function,
+    block: Option<BlockId>,
+    state: Option<StateOverride>,
+    provider: P,
+}
+
+impl<P> CallBuilder<P> {
+    pub(crate) fn new(provider: P, function: Function, input: Bytes) -> Self {
+        let request = CallRequest {
+            input: CallInput { input: Some(input), ..Default::default() },
+            ..Default::default()
+        };
+
+        Self { request, function, provider, block: None, state: None }
+    }
+
+    /// Sets the `from` field in the transaction to the provided value
+    pub const fn from(mut self, from: Address) -> Self {
+        self.request.from = Some(from);
+        self
+    }
+
+    /// Uses a Legacy transaction instead of an EIP-1559 one to execute the call
+    pub fn legacy(self) -> Self {
+        todo!()
+    }
+
+    /// Sets the `gas` field in the transaction to the provided value
+    pub const fn gas(mut self, gas: U256) -> Self {
+        self.request.gas = Some(gas);
+        self
+    }
+
+    /// Sets the `gas_price` field in the transaction to the provided value
+    /// If the internal transaction is an EIP-1559 one, then it sets both
+    /// `max_fee_per_gas` and `max_priority_fee_per_gas` to the same value
+    pub const fn gas_price(mut self, gas_price: U256) -> Self {
+        // todo: differentiate between eip-1559 and legacy once we have typedtx
```

Comment:
I'd say let's default to legacy txs while we don't have typedtx—it's what we use in tests IIRC

Reply:
Let's default to 1559, no?